### DISCUSSION TITLE
Getting started?

### DISCUSSION THREAD

**Discussion by Villhahallon** [BrianHumlicek/DSC-TLink#3] [General]
Hello, Was looking for a way to integrate my DSC panel with HA or something useful.
The other project don't seem to support my neo panel. I have an TL280(R)(E) that I would like to test this project out with but not so familiar with c# and vs studios project setup. 
Would it be possible to get some guidance with this to get started so I can experiment some with this?.

I manage to start the Demo project but I get some errors before getting anywhere.

**Comment by mtwo** [comment]
Someone on the Home Assistant forums [found](https://community.home-assistant.io/t/dsc-neo-integration/169078/32) an [official SDK for the TL280](https://www.dropbox.com/scl/fi/sexwwdd0xs7pj4y5s0cd7/InteractiveSDK_Release_2_0_1.zip?rlkey=10pj6h76zgfzgy2jagtvs1lmn&e=1&st=6zji65j1&dl=0). I haven't had time to try it out, but it might be pretty useful.

**Reply by BrianHumlicek** [reply]
Wow, that's awesome.  I'm going to pick through it now.

**Reply by BrianHumlicek** [reply]
That's an impressive haul.  Hopefully someone picks it up and runs with it as right now I don't have much spare bandwidth to do anything with it myself.  I have only looked at maybe 20% of it, but it seems like it has enough info to complete an integration.  I think it glosses over the underlying implementation which my TLink code might actually fill in.  This is an exciting find.

**Comment by BrianHumlicek** [comment]
@Villhahallon I'm happy to help get the project up and running.  What errors are you getting, and what have you tried to fix it?

Brian

**Reply by Villhahallon** [reply]
Great! as said i'm no expert on c# and vs studio...
I ran the code, edited the IP to fit my system. and get stuck at IListExtensions.cs line 59.
```c#
        public static byte TrimLeadingByte(this IList<byte> byteList)
        {
            byte result = byteList[0];
            byteList.RemoveAt(0);
            return result;
        }
```
`System.ArgumentOutOfRangeException: 'Index was out of range. Must be non-negative and less than the size of the collection. Arg_ParamName_Name'`

The byteList seem to return 0. 
![image](https://github.com/user-attachments/assets/ded30d58-8ccd-41c5-81ad-f58e36362771)
I'm not sure where to start looking as I don't have much information from this project.

I also tested a few ports but those could not connect so I'm sure something is connecting...

**Reply by BrianHumlicek** [reply]
That sounds like a parsing error because of an invalid byte in the payload.  It would be helpful to set a breakpoint at TLinkClient line 45 and see what is in the packet array before parsing.

If you are not sure how to do that maybe see if you can get the full stack trace so I can see where the parse is failing.

Also, the code for TrimLeadingByte is line 46, is that what you intended?

**Reply by Villhahallon** [reply]
Will try that tomorrow when I get back home. 

The result = byteList was at 59. But I will check again tomorrow.

**Reply by Villhahallon** [reply]
Sorry for late response. 
Line 45 say packet count 55
![image](https://github.com/user-attachments/assets/15780b5d-8b86-4d6f-b7c8-2a34d5c99822)
not sure if I got any useful data but I exported an csv of the packet.
[packetL45.csv](https://github.com/user-attachments/files/16478721/packetL45.csv)

**Comment by BrianHumlicek** [comment]
I loaded your data and found the issue is at TLinkSessionState line 155.  I made an assumption about the available data and yours doesn’t follow that assumption.  I won’t have time until later this week to research why I did that and what the appropriate fix is.  In the meantime you can probably comment out that line.

To be clear, the demo doesn’t do much other than parse the connection data and setup the encryption key for further communication.  Let me know if that gets you further along.

Brian

**Reply by Villhahallon** [reply]
I will try that fix for now.
Well it's a start to get the demo running.

I will write the results when I get time to test some more.

**Reply by BrianHumlicek** [reply]
I researched the issue, and it appears that the proper fix is to wrap the three assignment statements in the parseVarientData method with an if (length >=3).

`if (length >= 3)
{
    sessionState.MarketID = block.TrimLeadingByte();
    sessionState.ApprovalID = block.TrimLeadingByte();
    sessionState.CustomerID = block.TrimLeadingByte();
}`

**Reply by Villhahallon** [reply]
Thanks for taking your time to check.
It looks like the length is returning 2 so it passes that whole if statement.
Now I get stuck at the GSEncryptionKeyGenerator 204 `if ((inputBytes[index >> 3] & 1 << (index & 7)) != 0)`

Is there some settings I might need in the hardware for this to work? most of the config was set to use the original setup for the dsc system.

**Reply by BrianHumlicek** [reply]
I think the fix is to uncomment line 87 of TLinkSessionState (parseServiceRequestData).  I commented this to make it work on my system, and don't really understand why.  I will dig into it and figure out what the deal is.

As for settings, yes there are settings you need enabled to make it work, but I think if you are getting this far, you probably have it correct.

Brian

**Reply by BrianHumlicek** [reply]
So, the reason I needed to comment that out is there is another bug at line 99.  It has an errant +1; at the end.  It should be as follows.

int numberOfStatusBytes = payload.TrimLeadingByte();

Those two fixes to the parser should get you through.

I'll work on committing these updates soon.

Brian

**Reply by BrianHumlicek** [reply]
Try the latest commit.  It should incorporate all the latest fixes as well as a few minor refactors.

Brian

**Reply by Villhahallon** [reply]
Great!
I have tested the latest commit and I get no sudden errors when running the debugger. but the whole application exits after the terminal pops up. It loads all the dll files and then exits. `The program '[27468] Demo.exe' has exited with code 0 (0x0).`

I can see a lot of traffic going to and back from the dsc in my network.

**Reply by BrianHumlicek** [reply]
Well. this might be an anticlimactic finish, but that is all the demo does at the moment.  My intention was to initiate communication, parse the session state, and setup the AES keys for further communication.  It basically says "Hello" to the TL280 and that's it at the moment.

You can set breakpoints to examine the session state, and if you know what to say next to the TL280, you can use the TLinkClient to send and receive encrypted messages.

If nothing else, your effort uncovered a few latent bugs, and also verifies that it (seems to) works on a system other than mine.

Brian

**Reply by Villhahallon** [reply]
I was thinking that would be the case. Thank you for helping me with this. I will study the code some more and try what you said, hopefully i can get some information out from the system.

**Reply by pcanto-hopeit** [reply]
Great! I have also tested the latest commit on macOS and experienced similar behavior!

**Reply by mtwo** [reply]
Now that auth is working, it might be worth looking at the messages that the [official SDK](https://www.dropbox.com/scl/fi/sexwwdd0xs7pj4y5s0cd7/InteractiveSDK_Release_2_0_1.zip?rlkey=10pj6h76zgfzgy2jagtvs1lmn&e=1&st=6zji65j1&dl=0) can send out, either to recreate them in this codebase, or to attempt to use that SDK directly

**Comment by pcanto-hopeit** [comment]
Hello, I've been a bit tied up these past weeks and couldn't contribute as much, but I'm still very interested in network access to the DSC Neo. I've been looking for more information and found some details on https://github.com/rmilecki/dsc-itv2-protocol/discussions/2 that might help in deciphering the protocol. It references a file named ITv2Command.cs.txt. Unfortunately, I'm not a Windows user and don't have experience with Microsoft development. I'll keep an eye on your progress! Thank you for your efforts, and keep up the great work!

**Reply by BrianHumlicek** [reply]
I actually wrote that reply and posted that file.  Thanks for looking out though and keep any info coming.

So far, the sdk links referenced earlier in this thread hold the most promise I think.

Brian

**Comment by borisbirshtein** [comment]
Thanks for this! Really helpful.
